## Problem 17: Number Letter Counts

If the numbers 1 to 5 are written out in words: one, two, three, four, five, then there are $3 + 3 + 5 + 4 + 4 = 19$ letters used in total.

If all the numbers from 1 to 1000 (one thousand) inclusive were written out in words, how many letters would be used?


*NOTE*: Do not count spaces or hyphens. For example, 342 (three hundred and forty-two) contains 23 letters and 115 (one hundred and fifteen) contains 20 letters. The use of "and" when writing out numbers is in compliance with British usage.

In [1]:
import math

In [2]:
# Need numbers in english written form
digits = [
    '',    'one',    'two',    'three',    'four',     'five',    'six',     'seven',     'eight',    'nine',
    'ten', 'eleven', 'twelve', 'thirteen', 'fourteen', 'fifteen', 'sixteen', 'seventeen', 'eighteen', 'nineteen',
]

tens = [
    '', '', 'twenty', 'thirty', 'forty', 'fifty', 'sixty', 'seventy', 'eighty', 'ninety'
]

later = [
    '', 'thousand', 'million', 'billion', 'trillion', 'quadrillion', 'quintillion', 'sextillion', 'septillion', 'octillion', 'nonillion'
]

'''
Converts an integer number to its written english form.

eg.
>>> number_to_english(12345)
>>> 'twelvethousandthreehundredandfortyfive'
'''

def number_to_english(number : int) -> str:
    '''
    --- Function Description --------------------------------------------------------------------------------------------------
        Converts an integer to its written english form.
        
        Can do this using recursion:
        
        If number == : return ''
        
        If number < 20 : return 'one', 'two', etc.
        
        If number < 100 : return 'twenty', 'thirty', etc. + number_to_english(number % 10)
        
        Else number can be broken into the following XXX_YYY_ZZZ_...
            
            XXX = abc = a hundred and bc
            
            Need to isolate XXX:
                log10(XXX_YYY_ZZZ_...) = power of 10
                log10(XXX_YYY_ZZZ_...) // 3 = thousands / millions / etc.
                10 ** log10(XXX_YYY_ZZZ_...) // 3 * 3 = 1_000_000_...
                XXX_YYY_ZZZ_... // 10 ** log10(XXX_YYY_ZZZ_...) // 3 * 3 = XXX
            
            Need to remove XXX:
                10 ** log10(XXX_YYY_ZZZ_...) // 3 * 3 = 1_000_000_...
                XXX_YYY_ZZZ_... % 10 ** log10(XXX_YYY_ZZZ_...) // 3 * 3 = YYY_ZZZ_...
    ---------------------------------------------------------------------------------------------------------------------------
    
    --- Function Inputs -------------------------------------------------------------------------------------------------------
        : int : number : The number to be converted.
    ---------------------------------------------------------------------------------------------------------------------------
    
    --- Function Outputs ------------------------------------------------------------------------------------------------------
        : str : english : The number in written english form.
    ---------------------------------------------------------------------------------------------------------------------------
    
    --- Function Examples -----------------------------------------------------------------------------------------------------
        >>> number_to_english(12345)
        >>> 'twelvethousandthreehundredandfortyfive'
    ---------------------------------------------------------------------------------------------------------------------------
    '''
    
    # Check types of function inputs:
    if not isinstance(number, int): raise ValueError('Please enter an integer >= 0 for the number argument.')
    if number < 0: raise ValueError('Please enter an integer >= 0 for the number argument.')
    
    # Units
    if number < 20:
        return digits[number]
    
    # Tens
    if number < 100:
        return tens[number // 10] + number_to_english(number % 10)
    
    # Hundreds
    if number < 1000:
        if number % 100 == 0: return number_to_english(number // 100) + 'hundred'
        else: return number_to_english(number // 100) + 'hundred' + 'and' + number_to_english(number % 100)
    
    # All powers of 10 ** 3x
    if number < 10**33:
        return number_to_english(number // (10 ** (int(math.log10(number) // 3) * 3))) + later[(int(math.log10(number) // 3))] + number_to_english(number % 10 ** (int(math.log10(number) // 3) * 3))
    
    raise ValueError('Please enter an integer < 10**33 for the number argument.')

In [3]:
'''
Sums the letters in the written english form of all numbers up to and including max_number.

eg.
>>> count_letters(10)
>>> 39
'''

def count_letters(max_number : int) -> int:
    '''
    --- Function Description --------------------------------------------------------------------------------------------------
        Sums the letters in the written english form of all numbers up to and including max_number.
    ---------------------------------------------------------------------------------------------------------------------------
    
    --- Function Inputs -------------------------------------------------------------------------------------------------------
        : int : max_number : The max number to be included in the sum.
    ---------------------------------------------------------------------------------------------------------------------------
    
    --- Function Outputs ------------------------------------------------------------------------------------------------------
        : int : letter_sum : The sum of the letters of the numbers in written english form.
    ---------------------------------------------------------------------------------------------------------------------------
    
    --- Function Examples -----------------------------------------------------------------------------------------------------
        >>> count_letters(10)
        >>> 39
    ---------------------------------------------------------------------------------------------------------------------------
    '''
    
    # Check types of function inputs:
    if not isinstance(max_number, int): raise ValueError('Please enter an max_number >= 0 for the number argument.')
    if max_number < 0: raise ValueError('Please enter an integer >= 0 for the max_number argument.')
    
    letter_sum = 0
    for number in range(max_number+1):
        letter_sum += len(number_to_english(number))
    
    return letter_sum

In [4]:
max_number = 1000

solution = count_letters(max_number)

print(f'When written in their english form, the numbers 1 to {max_number:,} have {solution:,} letters combined.')

When written in their english form, the numbers 1 to 1,000 have 21,124 letters combined.


### Problem 17:

When written in their english form, the numbers 1 to 1,000 have 21,124 letters combined.